staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -327,7 +327,7 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
 	}
 	for _, obj := range objs {
 		if slice, ok := obj.(*resourceapi.ResourceSlice); ok {
-			slices = append(slices, slice)
+			slices = append(slices, slice.DeepCopy())
```

Comment:
@pohly @bart0sh 
ci still gives an error even though I use deepCopy in slices. This is weird. So I looked at the e2e code based on the error log.
```go
[sig-node] DRA [Feature:DynamicResourceAllocation] [NodeAlphaFeature:DynamicResourceAllocation] ResourceSlice [Serial] must be removed after plugin unregistration [sig-node, Feature:DynamicResourceAllocation, Serial]
k8s.io/kubernetes/test/e2e_node/dra_test.go:529
  STEP: Creating a kubernetes client @ 10/01/24 05:19:48.561
  STEP: Building a namespace api object, basename dra-node @ 10/01/24 05:19:48.561
  I1001 05:19:48.566828 1084 framework.go:275] Skipping waiting for service account
  STEP: start plugin and wait for ResourceSlice @ 10/01/24 05:19:48.659
  STEP: start Kubelet plugin @ 10/01/24 05:19:48.661
  I1001 05:19:48.661598    1084 draplugin.go:301] "Starting" logger="kubelet plugin test-driver.cdi.k8s.io" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:19:48.661773    1084 nonblockinggrpcserver.go:111] "GRPC server started" logger="kubelet plugin test-driver.cdi.k8s.io.dra" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:19:48.662152    1084 nonblockinggrpcserver.go:111] "GRPC server started" logger="kubelet plugin test-driver.cdi.k8s.io.registrar" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:19:48.662271    1084 resourceslicecontroller.go:259] "Starting ResourceSlice informer and waiting for it to sync" logger="kubelet plugin test-driver.cdi.k8s.io.ResourceSlice controller" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:19:48.662345    1084 reflector.go:305] Starting reflector *v1alpha3.ResourceSlice (10m0s) from k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265
  I1001 05:19:48.662361    1084 reflector.go:341] Listing and watching *v1alpha3.ResourceSlice from k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265
  I1001 05:19:48.664024    1084 reflector.go:368] Caches populated for *v1alpha3.ResourceSlice from k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265
  I1001 05:19:49.662878    1084 resourceslicecontroller.go:274] "ResourceSlice informer has synced" logger="kubelet plugin test-driver.cdi.k8s.io.ResourceSlice controller" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:19:49.662941    1084 resourceslicecontroller.go:124] "Starting" logger="kubelet plugin test-driver.cdi.k8s.io.ResourceSlice controller" node="tmp-node-e2e-44e8fec1-cos-stable-113-18244-151-88"
  I1001 05:27:09.665892    1084 reflector.go:871] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: Watch close - *v1alpha3.ResourceSlice total 8 items received
  I1001 05:29:48.664957    1084 reflector.go:389] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: forcing resync
  I1001 05:34:04.667663    1084 reflector.go:871] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: Watch close - *v1alpha3.ResourceSlice total 8 items received
  I1001 05:39:48.665446    1084 reflector.go:389] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: forcing resync
  I1001 05:40:19.669703    1084 reflector.go:871] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: Watch close - *v1alpha3.ResourceSlice total 7 items received
  I1001 05:48:07.671919    1084 reflector.go:871] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: Watch close - *v1alpha3.ResourceSlice total 8 items received
  I1001 05:49:48.665812    1084 reflector.go:389] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: forcing resync
  I1001 05:57:57.674440    1084 reflector.go:871] k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:265: Watch close - *v1alpha3.ResourceSlice total 10 items received
```
The error seems to occur here. I am curious that when the plugin is unregistration, we call the stop method, but we do not delete the resourceslices resource. Is this the cause of the error?
https://github.com/kubernetes/kubernetes/blob/22a30e7cbbee31b51da2f5e88833efbfad2c3f6b/test/e2e_node/dra_test.go#L529-L542

Reply:
https://github.com/kubernetes/kubernetes/blob/fbf1a0dc181ccbeb9925ad9c284d913a25c16562/pkg/kubelet/cm/dra/plugin/registration.go#L216 should remove the resource slices after the plugin has stopped.

Could it be that the controller doesn't stop properly anymore?

Reply:
This can be debugged by running locally similar to this (assuming that you have crio):

```console
$ CONTAINER_RUNTIME_ENDPOINT='unix:///var/run/crio/crio.sock' make test-e2e-node FOCUS="DynamicResourceAllocation.*must be removed after plugin unregistration" PARALLELISM=1 TEST_ARGS="--kubelet-flags=--cgroup-driver=systemd --feature-gates=DynamicResourceAllocation=true --service-feature-gates=DynamicResourceAllocation=true --runtime-config=api/all=true --kubelet-flags=--fail-swap-on=false --kubelet-flags=-v=5 --ginkgo.v --v=6" SKIP='xxxx' DBG=1
...
  I1001 14:11:47.801284  128127 resourceslicecontroller.go:274] "ResourceSlice informer has synced" logger="kubelet plugin test-driver.cdi.k8s.io.ResourceSlice controller" node="pohly-desktop"
  I1001 14:11:47.801368  128127 resourceslicecontroller.go:124] "Starting" logger="kubelet plugin test-driver.cdi.k8s.io.ResourceSlice controller" node="pohly-desktop"
  I1001 14:11:47.805403  128127 round_trippers.go:560] GET https://127.0.0.1:6443/api/v1/nodes/pohly-desktop 200 OK in 3 milliseconds
  I1001 14:11:47.806200  128127 round_trippers.go:560] GET https://127.0.0.1:6443/apis/resource.k8s.io/v1alpha3/resourceslices 200 OK in 4 milliseconds
  I1001 14:11:49.810813  128127 round_trippers.go:560] GET https://127.0.0.1:6443/apis/resource.k8s.io/v1alpha3/resourceslices 200 OK in 3 milliseconds
...
```

A status of what the `e2e_node.test` binary is waiting for can be requested with:
```console
$ sudo killall -USR1 e2e_node.test # runs as root
```

This shows:
```
You requested a progress report.
Currently there is no information about test progress available.
  ------------------------------
  You've requested a progress report:
    [sig-node] DRA [Feature:DynamicResourceAllocation] [NodeAlphaFeature:DynamicResourceAllocation] ResourceSlice [Serial] must be removed after plugin unregistration (Spec Runtime: 1m50.505s)
      /nvme/gopath/src/k8s.io/kubernetes/test/e2e_node/dra_test.go:529
      In [It] (Node Runtime: 1m50.492s)
        /nvme/gopath/src/k8s.io/kubernetes/test/e2e_node/dra_test.go:529
        At [By Step] start Kubelet plugin (Step Runtime: 1m50.481s)
          /nvme/gopath/src/k8s.io/kubernetes/test/e2e_node/dra_test.go:548

        Spec Goroutine
        goroutine 372 [select]
          github.com/onsi/gomega/internal.(*AsyncAssertion).match(0xc0002f6d90, {0x6267ca0, 0xc000b82d20}, 0x1, {0xc000e312c0, 0x1, 0x1})
            /nvme/gopath/src/k8s.io/kubernetes/vendor/github.com/onsi/gomega/internal/async_assertion.go:538
          github.com/onsi/gomega/internal.(*AsyncAssertion).Should(0xc0002f6d90, {0x6267ca0, 0xc000b82d20}, {0xc000e312c0, 0x1, 0x1})
            /nvme/gopath/src/k8s.io/kubernetes/vendor/github.com/onsi/gomega/internal/async_assertion.go:145
        > k8s.io/kubernetes/test/e2e_node.init.func22.4.4({0x7feaf67db008, 0xc000d46780})
            /nvme/gopath/src/k8s.io/kubernetes/test/e2e_node/dra_test.go:535
              | ginkgo.By("start plugin and wait for ResourceSlice")
              | kubeletPlugin := newKubeletPlugin(ctx, f.ClientSet, getNodeName(ctx, f), driverName)
              > gomega.Eventually(ctx, listResources).Should(matchNode, "ResourceSlice from kubelet plugin")
              | gomega.Consistently(ctx, listResources).WithTimeout(5*time.Second).Should(matchNode, "ResourceSlice from kubelet plugin")
              | 
          github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func2({0x6288d70, 0xc000d46780})
            /nvme/gopath/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/v2/internal/node.go:465
          github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
            /nvme/gopath/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:894
          github.com/onsi/ginkgo/v2/internal.(*Suite).runNode in goroutine 319
            /nvme/gopath/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:881

        Begin Additional Progress Reports >>
          ResourceSlice from kubelet plugin
          Expected
              <[]v1alpha3.ResourceSlice | len:0, cap:0>: 
                  []
          to consist of
              <[]*matchers.HaveFieldMatcher | len:1, cap:1>: [
                  {
                      Field: "Spec.NodeName",
                      Expected: <*matchers.EqualMatcher | 0xc000ea6ea0>{
                          Expected: <string>"pohly-desktop",
                      },
                      extractedField: nil,
                      expectedMatcher: nil,
                  },
              ]
          the missing elements were
              <[]*matchers.HaveFieldMatcher | len:1, cap:1>: [
                  {
                      Field: "Spec.NodeName",
                      Expected: <*matchers.EqualMatcher | 0xc000ea6ea0>{
                          Expected: <string>"pohly-desktop",
                      },
                      extractedField: nil,
                      expectedMatcher: nil,
                  },
              ]
        << End Additional Progress Reports
...

```

So it waits for the plugin to create one slice, which isn't happening.

Eventually (no pun intended, I swear) the `gomega.Eventually` should time out, but it's not clear when. In test/e2e, there is a shorter [default timeout](https://github.com/kubernetes/kubernetes/blob/fbf1a0dc181ccbeb9925ad9c284d913a25c16562/test/e2e/framework/test_context.go#L509) of 5 minutes, but in e2e_node it just keeps running - not sure why.

Time to use a debugger by adding `E2E_TEST_DEBUG_TOOL=dlv`.

The timeout gets set to 5 minutes, as expected. Not sure why that doesn't trigger.

I think I got it. The expectation is that a driver with no devices still publishes an empty resource slice, to indicate that it is running. The E2E node test relies on that because the fake driver isn't configured to have any devices (`testdriver.FileOperations{}`). The revised code figures out that the pool is empty and then doesn't create any slice. It needs to do that again.

Reply:
Looks like a Gomega bug. I filed https://github.com/onsi/gomega/issues/781.

Reply:
Thanks for the help, I will take a look at the problem when I have time in the next few days. It's a bit complicated...

Reply:
@klueska pointed out elsewhere that a driver might publish multiple pools, then remove all of them.

I think what we want is one slice per driver per node as indicator for "driver got started". We don't need one (empty!) slice per pool, that doesn't add much additional information.

Reply:
So do we need to modify the e2e code to not have empty resource slices or do we still need to meet this requirement?

Reply:
I tried running the e2e test (I finally learned how to do this), and still got the same error, but I can test it, so I should be able to fix this.
```bash
 [FAILED] An interrupt occurred and then the following failure was recorded in the interrupted node before it exited:
  Context was cancelled (cause: Interrupted by User) after 913.557s.
  ResourceSlice from kubelet plugin
  Expected
      <[]v1alpha3.ResourceSlice | len:0, cap:0>:
          []
  to consist of
      <[]*matchers.HaveFieldMatcher | len:1, cap:1>: [
          {
              Field: "Spec.NodeName",
              Expected: <*matchers.EqualMatcher | 0xc000783a40>{
                  Expected: <string>"vm-0-9-ubuntu",
              },
              extractedField: nil,
              expectedMatcher: nil,
          },
      ]
  the missing elements were
      <[]*matchers.HaveFieldMatcher | len:1, cap:1>: [
          {
              Field: "Spec.NodeName",
              Expected: <*matchers.EqualMatcher | 0xc000783a40>{
                  Expected: <string>"vm-0-9-ubuntu",
              },
              extractedField: nil,
              expectedMatcher: nil,
          },
      ]
  In [It] at: /root/go/src/k8s.io/kubernetes/test/e2e_node/dra_test.go:535 @ 10/17/24 22:18:01.694
```

Reply:
That we don't get that output as part of the job run is because of https://github.com/onsi/gomega/issues/781.

> So do we need to modify the e2e code to not have empty resource slices or do we still need to meet this requirement?

We need to meet the "exactly one empty slice" requirement when there are no devices.

Instead of testing this with `e2e_node` (hard to run, as you noticed), can you add a unit test case for starting with several different pools and then having only an empty slice at the end?

Reply:
Finally fixed it... :sweat_smile: Thanks Patrick :grin: